ReadMe.md
# Order Scrapping - iOS SDK 

# Contents

## Sub-Directories
- `src` : Source code for SDK
- `thirdparty` : Any thirdparty libraries/sources used in the project
- `docs` : Documents/Guides
- `scripts` : scripts like build-scripts for command line builds if any or similar.
- `bin` : SDK/library build (framework). This framework is a release artifact.
- `pocs` : Any PoC codes written for initial assessment of any feature/requirement
- `samples` : Sample iOS application(s) integrated with the SDK
- `tools` : Any tools needed for SDK and/or sample development. This can also include any external
  tools like code-static-analyser etc.

# Build Steps

Refer file `docs/ReadMe.SDK.build.md` for steps to build the SDK

# Testing the SDK

A sample application is developed and the source code for that is provided under `samples` directory.
Refer file `docs/ReadMe.app.md` file for steps to integrate SDK in any app.




docs/CHANGELOG.md
# BlackStraw OrderScrapper iOS SDK - Changelog

All notable changes to OrderScrapper iOS project will be documented in this file.

# Structure of this changelog:

- Header2 texts (sections marked with `##`) are in this format - `[Version] - YYYY-MM-DD`
- Header3 texts (sections marked with `###`) could be either one of these:
  - Added : This section to enlist the additions/implementations 
  - Changed : This section to enlist the changes to existing implementations
  - Notes : Any extra notes/remarks for the release


## [2.1.0] - 2022-06-21
 ### Added
 - RPA-1517  Added fix for the CSV scraping changes in amazon
 
## [2.0.9] - 2022-04-26
 ### Added
 - RPA-863 phase-2 changes completed
 
## [2.0.8] - 2022-03-31
 ### Added
 - Issue fixes added for the ticket RPA-1035
 
## [2.0.7] - 2022-03-10
 ### Added
 - Issue fixes added for the ticket RPA-1062

## [2.0.6] - 2022-02-24
 ### Added
 - Added online scraping for amazon for first phase release of RPA_863
 - Moved authentication scripts for all order sources  to server side (RPA- 833)
 - Issue fixes added for the tickets RPA- 773, RPA- 774, RPA- 775 , RPA-894
    
## [2.0.3] - 2022-02-04
  ### Added
   - Changes added for consistent connection flow of instacart and walmart (RPA-752)
   - Added scrape time for order details  in the backend DB (RPA-756)
   - Issue fixes added for MOSP-513, MOSP-514,MOSP-515, MOSP-516
   - Issue fixes added for MOSP-517, MOSP-518 and MOSP-531  
   - UI feedback to user during manual scraping (RPA-753 and RPA-585)
   
## [2.0.2] - 2022-01-19
  ### Added
   - Issue fix for user not able to connect instacart account (RPA-824)
   - Fixes added for the tickets RPA-689, RPA-692, RPA-710, RPA-711, RPA-713, RPA-743, RPA-755, RPA-757   
   
## [2.0.1] - 2022-01-03
  ### Added
   - bumped up the sdk version for multi-retailers features.

## [1.2.0] - 2021-12-15
  ### Added
   - Added new enum value connectionInProgress in AccountState enum. During connection flow on authentication completed updated account state as connectionInProgress  
   - In foreground scrapping on uploading all scrape orders updated account state as connected in DB and backend through API call
   
  ### Fixed
   - Issue fix for the account state changing to connectedButException in backgroundAuthentication (RPA-662)
   - Added ServicesStatusListener to handle the API down scenario from the SDK side. A callback given to the Application when the services are down.


## [1.1.9] - 2021-12-14
  ### Added
   - Sending input to API for switching scrapping type during manual scrapping/foreground scrapping
   - Instacart screen get visible for some time if we get any popup(like verification code and capcha) on screen
   - Instacart home screen get visible if we click on close or cancel button from the popup
   - Show error screen on getting failure for manual scrapping. Considering it will need UI changes for showing error screen. 

## [1.1.8] - 2021-12-10
  ### Added
   - Email validation updated which allows multiple dots after @ symbol
   - SDK provided an enum in the return value for the `fetchOrders()` method to the app stating the state of the scraping such as enqueued, in-progress etc. 
   - Static message updated on the Instacart login screen
   - Fixed added for the Instacart home screen get visible if we click on the close or cancel from the popup

## [1.1.7] - 2021-12-07
  ### Added
   - Added manual value in FetchRequestSource enum for manual scrapping. On receiving manual value as fetch request source in fetchOrders() from the app we are   doing foreground scrapping for all orders.
   - Labels and success messages updated on progress screen and success screen

## [1.1.6] - 2021-11-17

  ### Added
   - Added Walmart connect, disconnect and reconenct operations. Also added scrapping operations in both foreground and background
   - Added bug fixes related with the Walmart authentication
   - For Kroger added bug fixed in case of getting 'adblocker' error message during connection flow   
   - Added enhanced sentry logs which includes some additional attribute information and error logged as events 

## [1.1.5] - 2021-10-25
 
  ### Fixed
   - Optimized code to handle multiple network callbacks simultaneously ensuring the webview doesn't reload the same login URL multiple times
   - Removed the cool-off condition check in case the scraping is requested after notification click
   - Optimized core data code to be thread safe using a synchronized thread handler ensuring the object is not accessed simultaneously by different threads

## [1.1.4] - 2021-10-07
 
  ### Fixed
  - Sentry configs changes added such that if Sentry is enable in get_config API response then only Sentry will log the errors and events
  - Account state return as ConnectedButScrappingFailed to the app if account state is connected and scrapping is failing
  - Sentry configs changes added such that if Sentry is enable in get_config API response then only Sentry will log the errors and events

## [1.1.3] - 2021-09-17

  ### Fixed
  - On timeout update status API not getting called issue fixed 

## [1.1.2] - 2021-09-08

  ### Added
 - While scraping in background if user gets authentication challenge such as captcha, two-factor auth and approval auth, then maintained the scraping failure count in SDK and if in case the scraping failure count increases than the configured value then shown notification to user and on click of that notification navigated user to progress bar screen to handle auth challenge scenario.
 - Implemented configured cool-off period while scraping in background.
 - Implemented csv scraping in background mode if SDK gets configured scraping type as csv from backend.

## [1.1.1] - 2021-08-12
  
  ### Added
  - Static message on login screen for alerting the panelists for the csv file they will be sent after they connect their account.

## [1.1.0] - 2021-07-08

  ### Added

src/OrderScrapper/Views/LoginView.swift
//
//  LoginView.swift
//  OrderScrapper

import Foundation
import SwiftUI

struct LoginView : View {
    @State private var email: String = ""
    @State private var password: String = ""
    let bundle: Bundle! = Bundle(identifier: "ai.blackstraw.orderscrapper.OrderScrapper")
    @Environment(\.horizontalSizeClass) var sizeClass
    let padding_zero : CGFloat  = 0
    
    var body: some View {
        VStack {
            //Text Size
            Spacer(minLength: (sizeClass == .regular) ? 45 : 45)
            HStack {
                Image(IconNames.BackArrow, bundle: bundle)
                    .frame(width: (sizeClass == .regular) ? 30 : 23.25, height: (sizeClass == .regular) ? 30 : 23.25)
                    .padding(EdgeInsets(top: padding_zero, leading: (sizeClass == .regular) ? 20 : 20, bottom: padding_zero, trailing: (sizeClass == .regular) ? 16 : 16))
                Text(NSLocalizedString("heading_connect_amazon_account", tableName: nil, bundle: bundle, value: "", comment: ""))
                    .font(.system(size: (sizeClass == .regular) ? 22 : 17))
                    .foregroundColor(Color("heading_color", bundle: bundle))
                Spacer()
            }
            Spacer(minLength: (sizeClass == .regular) ? 35 : 35)
            VStack {
                HStack {
                    Text(NSLocalizedString("heading_please_sign_in_with_credentials", tableName: nil, bundle: bundle, value: "", comment: ""))
                        .font(.system(size: (sizeClass == .regular) ? 25 : 20))
                        .foregroundColor(Color("heading_color", bundle: bundle))
                    Spacer(minLength: (sizeClass == .regular) ? 25 : 25)
                }.padding(.top, (sizeClass == .regular) ? 50 : 30)
                .padding(.all, (sizeClass == .regular) ? 25 : 15)
                
                //Error view
                HStack {
                    Image(IconNames.Error, bundle: bundle)
                        .frame(width: (sizeClass == .regular) ? 35 : 35, height: (sizeClass == .regular) ? 35 : 35, alignment: .leading)
                    Text(NSLocalizedString("error_enter_valid_username_password", tableName: nil, bundle: bundle, value: "", comment: ""))
                        .font(.system(size: (sizeClass == .regular) ? 17 : 12))
                        .padding(EdgeInsets(top: padding_zero, leading: padding_zero, bottom: padding_zero, trailing: (sizeClass == .regular) ? 16 : 16))
                        .foregroundColor(Color("error_color", bundle: bundle))
                    Spacer()
                }.padding(EdgeInsets(top: padding_zero, leading: (sizeClass == .regular) ? 21 : 16, bottom: padding_zero, trailing: (sizeClass == .regular) ? 21 : 16))
                .hidden()
                HStack {
                    Text(NSLocalizedString("label_email_or_mobile_number", tableName: nil, bundle: bundle, value: "", comment: ""))
                        .font(.system(size: (sizeClass == .regular) ? 19 : 14))
                        .foregroundColor(Color("label_color", bundle: bundle))
                    Spacer()
                }.padding(EdgeInsets(top: (sizeClass == .regular) ? 15 : 10, leading: (sizeClass == .regular) ? 21 : 16, bottom: padding_zero, trailing: (sizeClass == .regular) ? 21 : 16))
                
                TextField(NSLocalizedString("label_email_or_mobile_number", tableName: nil, bundle: bundle, value: "", comment: ""), text: $email)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                    .border(Color("border_color_text_field", bundle: bundle), width: (sizeClass == .regular) ? 1 : 1)
                    .cornerRadius((sizeClass == .regular) ? 4 : 4)
                    .padding(EdgeInsets(top: (sizeClass == .regular) ? 10 : 5, leading: (sizeClass == .regular) ? 21 : 16, bottom: padding_zero, trailing: (sizeClass == .regular) ? 21 : 16))
                    .frame(width: .infinity , height: (sizeClass == .regular) ? 60 : 40)
                
                HStack {
                    Text(NSLocalizedString("label_password", tableName: nil, bundle: bundle, value: "", comment: ""))
                        .font(.system(size: (sizeClass == .regular) ? 19 : 14))
                        .foregroundColor(Color("label_color", bundle: bundle))
                    Spacer()
                }.padding(EdgeInsets(top: padding_zero, leading: (sizeClass == .regular) ? 21 : 16, bottom: padding_zero, trailing: (sizeClass == .regular) ? 21 : 16))
                
                SecureField(NSLocalizedString("label_password", tableName: nil, bundle: bundle, value: "", comment: ""), text: $password)
                    .textFieldStyle(RoundedBorderTextFieldStyle())
                    .border(Color("border_color_text_field", bundle: bundle), width: (sizeClass == .regular) ? 1 : 1)
                    .cornerRadius((sizeClass == .regular) ? 4 : 4)
                    .padding(EdgeInsets(top: (sizeClass == .regular) ? 10 : 5, leading: (sizeClass == .regular) ? 21 : 16, bottom: padding_zero, trailing: (sizeClass == .regular) ? 21 : 16))
                    .frame(width: .infinity , height: (sizeClass == .regular) ? 60 : 40)
                
                //SignIn Button
                HStack {
                    Spacer()
                    Button(action: {
                        print("Submit tapped!")
                    }) {
                        HStack(alignment: .center) {
                            Text(NSLocalizedString("btn_submit", tableName: nil, bundle: bundle, value: "", comment: ""))
                                .fontWeight(.semibold)
                                .font(.system(size: (sizeClass == .regular) ? 23 : 18))
                                .padding(EdgeInsets(top: padding_zero, leading: (sizeClass == .regular) ? 10 : 10, bottom: padding_zero, trailing: padding_zero))
                                .foregroundColor(Color("btn_color", bundle: bundle))
                            Spacer()
                            Image(IconNames.RightArrow, bundle: bundle)
                                .frame(width: (sizeClass == .regular) ? 18 : 18, height:(sizeClass == .regular) ? 17 : 17)
                                .padding(EdgeInsets(top: padding_zero, leading: (sizeClass == .regular) ? 10 : 10, bottom: padding_zero, trailing: (sizeClass == .regular) ? 23 : 21))
                        }
                        .frame(width: (sizeClass == .regular) ? 150 : 133 , height: (sizeClass == .regular) ? 55 : 48, alignment: .leading)
                        .padding(EdgeInsets(top: padding_zero, leading: (sizeClass == .regular) ? 23 : 21, bottom: padding_zero, trailing: padding_zero))
                        .foregroundColor(.white)
                        .background(LinearGradient(gradient: Gradient(colors: [Color("color_linear_gradient2", bundle: bundle), Color("color_linear_gradient1", bundle: bundle)]), startPoint: .leading, endPoint: .trailing))
                        .cornerRadius((sizeClass == .regular) ? 24 : 24)
                    }.padding(EdgeInsets(top: (sizeClass == .regular) ? 60 : 45, leading: padding_zero, bottom: padding_zero, trailing: (sizeClass == .regular) ? 21 : 16))
                }
                Spacer()
            }.background(Color.white)
            .cornerRadius((sizeClass == .regular) ? 35 : 35, corners: [.topLeft, .topRight])
        }.background(RadialGradient(gradient: Gradient(colors: [Color("color_radial_gradient1", bundle: bundle), Color("color_radial_gradient2", bundle: bundle)]), center: .center, startRadius: 1, endRadius: 100))
        .edgesIgnoringSafeArea(.all)
        
    }
}


struct LoginView_Previews: PreviewProvider {
    static var previews: some View {
        Group {
            LoginView()
                .previewDevice("iPhone 12 mini")
                .previewDisplayName("iPhone 12 mini")
            LoginView()
                .previewDevice("iPhone 12 Pro Max")
                .previewDisplayName("iPhone 12 Pro Max")
            LoginView()
                .previewDevice("iPhone 8")
                .previewDisplayName("iPhone 8")
            LoginView()
                .previewDevice("iPhone SE (2nd generation)")
                .previewDisplayName("iPhone SE (2nd generation)")
            LoginView()
                .previewDevice("iPad Pro (12.9-inch) (4th generation)")
                .previewDisplayName("iPad Pro(11-inch)(2nd generation)")
        }
    }
}

